File: sdk/typescript/src/Bip32.ts
import { PrivateKey } from './CryptoTypes.js';
const Mnemonic = require('bitcore-mnemonic') as any;

import crypto from 'crypto';

/**
 * Representation of a BIP32 node.
 */
export class Bip32Node {
	privateKey: PrivateKey;
	chainCode: Buffer;
	/**
	 * Creates a BIP32 node around a key and data.
	 * @param {Uint8Array} hmacKey BIP32 HMAC key.
	 * @param {Uint8Array} data BIP32 seed.
	 */
	constructor(hmacKey: Uint8Array, data: Uint8Array) {
		const hmac = crypto.createHmac('sha512', hmacKey);
		hmac.update(data);
		const hmacResult = hmac.digest();

		/**
		 * Private key associated with this node.
		 * @type PrivateKey
		 */
		this.privateKey = new PrivateKey(hmacResult.subarray(0, PrivateKey.SIZE));

		/**
		 * Chain code associated with this node.
		 * @type Uint8Array
		 */
		this.chainCode = hmacResult.subarray(PrivateKey.SIZE);
	}

	/**
	 * Derives a direct child node with specified identifier.
	 * @param {number} identifier Child identifier.
	 * @returns {Bip32Node} BIP32 child node.
	 */
	deriveOne(identifier: number): Bip32Node {
		const childData = new Uint8Array(1 + PrivateKey.SIZE + 4);
		childData[0] = 0;
		childData[childData.length - 4] = 0x80;

		for (let i = 0; 4 > i; ++i)
			childData[childData.length - 1 - i] |= (identifier >> (8 * i)) & 0xFF;

		for (let i = 0; i < PrivateKey.SIZE; ++i)
			childData[1 + i] = this.privateKey.bytes[i];

		return new Bip32Node(this.chainCode, childData);
	}

	/**
	 * Derives a descendent node with specified path.
	 * @param {Array<number>} path BIP32 path.
	 * @returns {Bip32Node} BIP32 node at the end of the path.
	 */
	derivePath(path: Array<number>): Bip32Node {
		/** @type {Bip32Node} */
		let nextNode: Bip32Node = this;
		path.forEach(identifier => {
			nextNode = nextNode.deriveOne(identifier);
		});

		return nextNode;
	}
}

/**
 * Factory of BIP32 root nodes.
 */
export class Bip32 {
	private _rootHmacKey: Buffer;
	private _mnemonicLanguage: string;
	/**
	 * Creates a BIP32 root node factory.
	 * @param {string} curveName Elliptic curve to use.
	 * @param {string} mnemonicLanguage Language of constructed mnemonics.
	 */
	constructor(curveName = 'ed25519', mnemonicLanguage = 'english') {
		/**
		 * @private
		 */
		this._rootHmacKey = Buffer.from(`${curveName} seed`);

		/**
		 * @private
		 */
		this._mnemonicLanguage = mnemonicLanguage;
	}

	/**
	 * Creates a BIP32 root node from a seed.
	 * @param {Uint8Array} seed BIP32 seed.
	 * @returns {Bip32Node} BIP32 root node.
	 */
	fromSeed(seed: Uint8Array): Bip32Node {
		return new Bip32Node(this._rootHmacKey, seed);
	}

	/**
	 * Creates a BIP32 root node from a BIP39 mnemonic and password.
	 * @param {string} mnemonic BIP32 mnemonic.
	 * @param {string} password BIP32 mnemonic password.
	 * @returns {Bip32Node} BIP32 root node.
	 */
	fromMnemonic(mnemonic: string, password: string): Bip32Node {
		const wordlist = Mnemonic.Words[this._mnemonicLanguage.toUpperCase()];
		return this.fromSeed(new Mnemonic(mnemonic, wordlist).toSeed(password));
	}

	/**
	 * Creates a random BIP32 mnemonic.
	 * @param {number} seedLength Length of random seed to use when generating mnemonic.
	 * @returns {string} Random mnemonic created with the specified entropy.
	 */
	random(seedLength: number = 32): string {
		const wordlist = Mnemonic.Words[this._mnemonicLanguage.toUpperCase()];
		return new Mnemonic(seedLength * 8, wordlist).phrase;
	}
}

File: sdk/dart/lib/Bip32.dart
import 'dart:typed_data';
import 'package:bip39_mnemonic/bip39_mnemonic.dart' as bip39;
import 'package:pointycastle/export.dart' as pointy;

import 'CryptoTypes.dart';

class Bip32Node {
  late PrivateKey privateKey;
  late Uint8List chainCode;

  Bip32Node(Uint8List hmacKey, Uint8List data) {
    var hmac = pointy.HMac(pointy.SHA512Digest(), 128)
      ..init(pointy.KeyParameter(hmacKey));
    var hmacResult = hmac.process(data);
    privateKey = PrivateKey(hmacResult.sublist(0, 32));
    chainCode = Uint8List.fromList(hmacResult.sublist(32));
  }

  Bip32Node deriveOne(int identifier) {
    var childData = Uint8List(1 + 32 + 4);
    childData[0] = 0;
    childData[childData.length - 4] = 0x80;

    for (var i = 0; 4 > i; ++i) {
      childData[childData.length - 1 - i] |= (identifier >> (8 * i)) & 0xFF;
    }

    for (var i = 0; i < 32; ++i) {
      childData[1 + i] = privateKey.bytes[i];
    }

    return Bip32Node(chainCode, childData);
  }

  Bip32Node derivePath(List<int> path) {
    var nextNode = this;
    path.forEach((identifier) {
      nextNode = nextNode.deriveOne(identifier);
    });

    return nextNode;
  }
}

class Bip32 {
  late Uint8List rootHmacKey;
  late bip39.Language mnemonicLanguage;

  Bip32(
      [String curveName = 'ed25519',
      bip39.Language mnemonicLanguage = bip39.Language.english]) {
    rootHmacKey = Uint8List.fromList('$curveName seed'.codeUnits);
    this.mnemonicLanguage = mnemonicLanguage;
  }

  Bip32Node fromSeed(Uint8List seed) {
    return Bip32Node(rootHmacKey, seed);
  }

  Bip32Node fromMnemonic(String mnemonic, String password) {
    return fromSeed(Uint8List.fromList(bip39.Mnemonic.fromSentence(
            mnemonic, mnemonicLanguage,
            passphrase: password)
        .seed));
  }

  String random([int seedLength = 32]) {
    var wordList = bip39.Mnemonic.generate(mnemonicLanguage,
        passphrase: '', entropyLength: seedLength * 8);
    return bip39.Mnemonic(wordList.entropy, mnemonicLanguage).sentence;
  }
}
